### ISSUE TITLE
What causes client disconnetion?

### ISSUE THREAD

**Issue by akreienbring** [primus/primus#389]
Hello,
I'm using the websockets transformer and configured the primus server with the option timeout: false.
Ideally I want my clients (an app based on android 4.4.2 / 5.1) to stay connected all the time to be able to receive messages.
But every once and a while the sparks are disconnecting and the immediatly reconnecting. 
This isn't happening because the app was killed by the OS. The app is running fine all the time.

Could it be that networking is disabled when the app is in Background mode? I tried to prevent this with this plugin https://github.com/katzer/cordova-plugin-background-mode but it does not help.

Could it be that the cpu stops working when the OS goes into deep sleep? I acquiered a partial wakelock with this plugin https://github.com/Viras-/cordova-plugin-powermanagement but... still the same reoccuring dis- / reconnection cycle...

Maybe its some sort of network delay or outtage, e.g. the change of cellphone network areas... But I'm not aware of any client side timeout... or is some kind of missing pong or heartbeat killing my client?

Any Ideas?

**Comment by lpinca** [comment]
I don't think that the disconnections are caused by missing hearthbeats, you disabled them on primus and you are running the `websockets` transformer which afaik doesn't have a heartbeat system.

Do the disconnections happen when the app is in foreground?

**Comment by 3rd-Eden** [comment]
As we're talking about cordova here, I assume that this is on mobile phones. Couldn't it just that your user is navigating between once wifi hotspot to another causing brief downtime which triggers a reconnect?

Also, make sure that you've actually re-compiled the client code after you've changed your setting as these settings a written in to the client file. 

And backgrounding applications can also cause this as it does kill things such as `setTimeout`

**Comment by akreienbring** [comment]
Hi, thanks for answering! I'm still examining this.

@3rd-Eden: No, change of wifi can not be the reason, because both of the phones / clients in question are mine and they don't travel between hotspots. Also it has nothing to do with compiling as I'm setting timeout: false for a very long time now and restarted the server and the clients very often since then....

@lpinca: regarding your question: hard to say.. as the disconnections occure within rather long time distances. Let't say between 10 to 40 minutes... But I will try to keep the phone in foreground and awake for another test..

What I can tell so far: It SEEMS, that the disconnections only occur when I'm connected to the server via the GSM Network. As long as I use my WIFI at home everything SEEMS to work fine. It will need further inspections to verify that.... 

But still the important question is: What could make the client disconnect (trigger a disconnection event on the server) when the server timeout is false and the app is running?

**Comment by lpinca** [comment]
> But still the important question is: What could make the client disconnect (trigger a disconnection event on the server) when the server timeout is false and the app is running?

Whatever makes a TCP connection close.

**Comment by 3rd-Eden** [comment]
@akreienbring On more question regarding this, on which platform is your application hosted?

**Comment by akreienbring** [comment]
On the server side: node.js 0.12.7 (Test: Windows, hosted: Linux i guess)
On the client side: Andrdroid (Kitkat & Lollipop)

I tested (nearly) all transformers (websockets, sockjs, engine.io, faye). All showed frequent disconnections. Sockjs a little bit less than the others. Disconnect / reconnect cycles occured every 2 min. at the shortest to 40 min. at the longest.

If connected via WLAN there are less and if connected via cell phone network there are more disconnections.

They occure when the app goes to background or the phone goes to sleep mode.

Independet from the app state primus keeps reconnecting immediatly (Which is good!) 

I observed situations on the server where the client just seems to connect again (different spark), without disconnecting before.

**Comment by henryleu** [comment]
I have the same problem. disconnection and reconnecting happens each 35ms.  I searched primus source code and failed to find 35000. it is so frustrating of me.
